File: ds_1/day_1/maximum-subarray.py
# https://leetcode.com/problems/maximum-subarray/

from typing import List

class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        if len(nums) == 0:
            return nums[0]
        
        maxSum = nums[0]
        curSum = nums[0]

        for num in nums[1:]:
            curSum = max(num, curSum + num)
            if curSum > maxSum:
                maxSum = curSum
            if curSum < 0:
                curSum = 0
        return maxSum

File: 2022/easy/maximum-subarray.ts
// https://leetcode.com/problems/maximum-subarray/

function maxSubArray(nums: number[]): number {
    let maxSum = nums[0]
    let curSum = nums[0]

    for (let i = 1; i < nums.length; i ++ ) {
        if (nums[i] > curSum + nums[i]) {
            curSum = nums[i]
        } else {
            curSum += nums[i]
        }
        if (curSum > maxSum) {
            maxSum = curSum
        }
    }

    return maxSum;
};


console.log(maxSubArray([-2,1,-3,4,-1,2,1,-5,4]))
console.log(maxSubArray([1, 2]))